src/charm.py

```diff
@@ -3,61 +3,77 @@
+
+class ServeWebsiteEvent(EventBase):
+    pass
+
+
+class GitLabK8sCharmEvents(CharmEvents):
+    configure_pod = EventSource(ConfigurePodEvent)
+    serve_website = EventSource(ServeWebsiteEvent)
+
+
+class NotLeaderError(Exception):
+    def __init__(self):
+        super().__init__('not leader')
+        self.status = WaitingStatus('Not leader')
 
 
 class GitLabK8sCharm(CharmBase):
     state = StoredState()
+    on = GitLabK8sCharmEvents()
 
     def __init__(self, framework, key):
         super().__init__(framework, key)
 
-        for event in (self.on.start,
-                      self.on.upgrade_charm,
-                      self.on.config_changed,
-                      self.on.mysql_relation_changed):
-            self.framework.observe(event, self.on_start)
-        self.framework.observe(self.on.website_relation_joined, self)
-
-        self.website = HTTPInterfaceProvides(self, 'website')
-        self.mysql = MySQLInterfaceRequires(self, 'mysql')
+        self.website = HTTPServer(self, 'website')
+        self.mysql = MySQLServer(self, 'mysql')
         self.gitlab_image = OCIImageResource(self, 'gitlab_image')
 
+        self.state.is_started = getattr(self.state, 'is_started', False)
+        self.state.has_website = getattr(self.state, 'has_website', False)
 
-    def on_install(self, event):
-        self.state.is_started = False
+        self.framework.observe(self.on.start, self.check_ready)  # TODO: change to install
+        self.framework.observe(self.on.leader_elected, self.check_ready)
+        self.framework.observe(self.on.upgrade_charm, self.check_ready)
+        self.framework.observe(self.on.config_changed, self.check_ready)
+        self.framework.observe(self.mysql.on.database_available, self.check_ready)
+        self.framework.observe(self.mysql.on.database_changed, self.check_ready)
+        self.framework.observe(self.on.configure_pod, self.configure_pod)
+        self.framework.observe(self.website.on.new_client, self.serve_website)
 
-    def on_start(self, event):
-        unit = self.framework.model.unit
-        if not self.gitlab_image.fetch():
-            unit.status = BlockedStatus('Missing or invalid image resource')
-            return
-        if not self.mysql.is_joined:
-            unit.status = BlockedStatus('Missing database')
-        if not self.mysql.is_single:
-            unit.status = BlockedStatus('Too many databases')
-        if not self.mysql.is_ready:
-            unit.status = WaitingStatus('Waiting for database')
-            return
-        if not self.framework.model.unit.is_leader():
-            unit.status = WaitingStatus('Not leader')
-            return
-        unit.status = MaintenanceStatus('Configuring container')
-        self.framework.model.pod.set_spec({
+    def check_ready(self, event):
+        try:
+            self.mysql.database()
+            self.gitlab_image.fetch()
+            if not self.model.unit.is_leader():
+                raise NotLeaderError()
```

Comment:
That's an awkward pattern. If we continue down that road, we'll have here every single method call that the code ever needs which might indicate the whole thing needs to wait for everything else. It's still following the original mindset of checking the entire world upfront before trying anything at all. Instead, would be nice if we could try being a bit more traditional in that you only need to watch for cases that you care about when you care about them.

Reply:
Honestly, I don't see a way around it.  All three of these things are hard requirements that have to be checked and met before the pod spec can be created, so there inherently has to be a place in the code that brings them together.  Even if we have event handlers that watch for each of these separately and set a stored state, then we still need some place that checks all three of those stored state flags and synthesizes them together into the combined "ready" event.  But maybe that at least will be easier to read, so I'll try that approach.

Reply:
I moved as much as possible to individual event handlers, but there was still some gate checking required at the top of `configure_pod`.  I'd like review on the new pattern and to discuss further if there's something more fundamental I'm missing.